src/funkin/vis/dsp/SpectralAnalyzer.hx

```diff
@@ -1,476 +1,125 @@
+
+	public function new(audioSource:lime.media.AudioSource, barCount:Int, maxDelta:Float = 0.01, peakHold:Int = 30)
+	{
+		this.audioSource = audioSource;
+		this.barCount = barCount;
         this.maxDelta = maxDelta;
         this.peakHold = peakHold;
-        this.barCount = barCount;
-
-        #if web
-        htmlAnalyzer = new AnalyzerNode(cast audioClip);
-        #end
-        // maxFreq = Math.min(maxFreq, audioClip.audioBuffer.sampleRate / 2);
 
-        calcBars(barCount, peakHold);
-
-        resizeBlackmanWindow(fftN);
-    }
-
-    function resizeBlackmanWindow(size:Int)
-    {
-        if (blackmanWindow.length == size) return;
-        blackmanWindow.resize(size);
-        for (i in 0...size) {
-            blackmanWindow[i] = calculateBlackmanWindow(i, size);
+        blackmanWindow.resize(fftN);
+        for (i in 0...fftN) {
+            blackmanWindow[i] = calculateBlackmanWindow(i, fftN);
         }
-    }
-
-    static inline function clamp(val:Float, min:Float, max:Float):Float
-    {
-        return val <= min ? min : val >= max ? max : val;
-    }
-
-
-    // For second stage, make this return a second set of recent peaks
-    public function getLevels(debugMode:Bool, ?elapsed:Float):Array<Bar>
-    {   
-        
-        var levels = new Array<Bar>();
-
-        var index:Int = audioClip.currentFrame ?? 0;
-        var indices:Array<Int> = [index];
-        var halfStride:Int = Std.int(fftN / 2);
-        var stride:Int = Std.int(fftN * smoothing);
-        if (index - halfStride > 0) indices = [index - halfStride];
-        // if (audioClip.audioBuffer.data.length > index + halfStride) indices = [index + stride];
-        var amplitudesSet = new Array<Array<Float>>();
-        var webAmplitudes:Array<Float> = [];
-
-        var prevLevels:Array<Float> = previousSTFT;
-        var sameAsPrev:Bool = false;
-
-        for (index in indices) {
-            #if web
-            var amplitudes:Array<Float> = htmlAnalyzer.getFloatFrequencyData();
-            #else
-            var amplitudes = stft(index, elapsed);
-            #end
-            
-            // sameAsPrev = amplitudes == prevLevels;
-
-            amplitudesSet.push(amplitudes);
-            webAmplitudes = amplitudes;
-
+	}
+
+	public function getLevels():Array<Bar>
+	{
+        var numOctets = Std.int(audioSource.buffer.bitsPerSample / 8);
+		var wantedLength = fftN * numOctets * numChannels;
+		var startFrame = currentFrame;
+        startFrame -= startFrame % numOctets;
+		var segment = audioSource.buffer.data.subarray(startFrame, min(startFrame + wantedLength, audioSource.buffer.data.length));
+		var signal = segment.toInterleaved(audioSource.buffer.bitsPerSample);
+
+		if (numChannels > 1) {
+			var mixed = new Array<Float>();
+			mixed.resize(Std.int(signal.length / numChannels));
+			for (i in 0...mixed.length) {
+				mixed[i] = 0.0;
+				for (c in 0...numChannels) {
+					mixed[i] += 0.7 * signal[i*numChannels+c];
+                    break;
```

Comment:
Whats with this break?

Reply:
That was left in by mistake! I'll fix that.

Reply:
Okay I fixed that. I meant to just temporarily only use the left channel while debugging the weird data tear issue that was caused by `currentFrame` sometimes being an odd number. And I can confirm removing the break didn't break anything.